### PR TITLE
[types] Language storage refactor

### PR DESCRIPTION
**Pull request by runtian-zhou**
## Motivation

Refactoring move specific concepts from `libra_types` to `move_core_types`. Here's a list of items that are moved:
1. TypeTag, StructTag, these are the fully qualified name for a move type/resource.
2. ModuleId, the fully qualified name for a module
3. MoveResource, the rust trait to auto generate code to fetch resource from storage.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Related PRs

#3789

### CODE REVIEW

**Comment by jolestar** [comment]
Should AccessPath and ContractEvent also be part of the move-core-type as well?

**Comment by dariorussi** [comment]
> Should AccessPath and ContractEvent also be part of the move-core-type as well?

they will, or better ContractEvent for sure, AccessPath not sure, @runtian-zhou?

File: types/src/language_storage.rs [hunk 1]
```diff
@@ -1,144 +1 @@
-// Copyright (c) The Libra Core Contributors
-// SPDX-License-Identifier: Apache-2.0
```
Comment (dariorussi) [inline types/src/language_storage.rs:1]: is this file gone? I think it's missing, given the compiler error....
also I wonder if `language_storage` is a good name for this, though something to think about and do in time in another PR

**Comment by runtian-zhou** [comment]
> Should AccessPath and ContractEvent also be part of the move-core-type as well?

That's a really great question! I don't have any further plans on them yet but lemme share my thoughts on those two:
1. For AccessPath, I feel like this is a more libra specific concept, as for move really all we need is to specify the hight level api of how things are stored. In the current code, Move expects two key value storage that stores module and resources with key type ModuleID and StructTag. It just happens that Libra merged the two storage together as an AccessPath. Thus we probably should still keep AccessPath where it is.
2. For ContractEvent, I'm really not sure. The reason is that Event is somewhat a concept more related to libra rather than the move bytecode itself. It is currently implemented as a native function and could possibly refactored out from the core semantics of move bytecode.

**Comment by jolestar** [comment]
@runtian-zhou  Thanks, I got it. I think move-core-types & move-vm should not dependency libra-types in the final, but the ChainState trait need AccessPath & ContractEvent, so they may be more appropriate concepts of Move.

**Comment by runtian-zhou** [comment]
> @runtian-zhou  Thanks, I got it. I think move-core-types & move-vm should not dependency libra-types in the final, but the ChainState trait need AccessPath & ContractEvent, so they may be more appropriate concepts of Move.

That’s exactly where we are planning to go in the future. I think it should be fairly easy to get rid of AccessPath in MoveVM. It’s a bit harder to move ContractEvent right now and probably need some amount of refactor to move native function implementation to another crate.
